testsuite/tests/typing-gadts/pr5985.ml

```diff
@@ -11,8 +11,9 @@ end;; (* fail *)
 Line 3, characters 2-29:
 3 |   type _ t = T : 'a -> 'a s t
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
-Error: In this definition, a type variable cannot be deduced
-       from the type parameters.
+Error: In the expanded definition
+         type 'b t = private T : 'a -> 'c t constraint 'b = 'a s
```

Comment:
I find the expansion fairly surprising here, for two reasons:
- I don't understand where `private` is coming from (my memory of private datatype declarations is rusty, I'm not sure they actually exist and I suspect that if they do the "private" type shown is not equivalent to the original declaration)
- I don't think that the type equation introduced by the GADT constructor is in fact equivalent to the `constraint` form that is shown. Obviously when you have several constructors it is not possible to translate different GADT equations into a single toplevel constraint, but even with a single constructor the type-checking rules are different. (I think a constraint is enforced all the time as well-formedness, while GADTs are checked dynamically upon opening the constructor in pattern-matching.) Lastly, when I input this "expanded definition" in the toplevel it rejects it -- because indeed constraints and GADTs don't mix nicely.

I suspect that this is none of your doing and that the type printer is just behaving weirdly, but the result looks very strange to me. If we "reprint" something the user wrote, I would expect the new thing to be equivalent to the user thing, and this is not what is going on here.

Reply:
> I suspect that this is none of your doing
Yeah, what I'm doing is printing out the internal state of the declaration at the time the variance check fails -- there's (almost) no change of what the type checker is giving me before passing it to the printing function.
I also tried some variations, such as just passing the original definition of the type (the one that is passed to the check function), but that doesn't work for two reasons: 1) The type variables are only named properly during the check, so we don't have usable names, and 2) some expansion is definitely required in the case the type variables aren't directly obvious from the definition.

> Lastly, when I input this "expanded definition" in the toplevel it rejects it
This might be a more fundamental problem, if the internal state of the typechecker is more expressive than the syntax, this can't be avoided. I think if it's obvious what is implied (which is not exactly the case here, so it needs to be improved), we might be fine with a sort of pseudo-syntax for the error message.

I can try to trace some of the weirdness (I think `private` might be the easiest to follow, since I guess it's just a bool/variant being changed) to see what kind of thing we're dealing with here.